compiler/sem/sighashes.nim

```diff
@@ -164,28 +164,18 @@ proc hashType(c: var MD5Context, t: PType; flags: set[ConsiderFlag]) =
       else:
         c.hashSym(t.sym)
 
-      var symWithFlags: PSym
-      template hasFlag(sym): bool =
-        let ret = {sfAnon, sfGenSym} * sym.flags != {}
-        if ret: symWithFlags = sym
-        ret
-      if hasFlag(t.sym) or (t.kind == tyObject and t.owner.kind == skType and t.owner.typ.kind == tyRef and hasFlag(t.owner)):
-        # for `PFoo:ObjectType`, arising from `type PFoo = ref object`
-        # Generated object names can be identical, so we need to
-        # disambiguate furthermore by hashing the field types and names.
-        if t.n.len > 0:
-          let oldFlags = symWithFlags.flags
-          # Hack to prevent endless recursion
-          # xxx instead, use a hash table to indicate we've already visited a type, which
-          # would also be more efficient.
-          symWithFlags.flags.excl {sfAnon, sfGenSym}
-          hashTree(c, t.n, flags + {CoHashTypeInsideNode})
-          symWithFlags.flags = oldFlags
-        else:
-          # The object has no fields: we _must_ add something here in order to
-          # make the hash different from the one we produce by hashing only the
-          # type name.
-          c &= ".empty"
+      if t.sym.flags * {sfAnon, sfGenSym} != {} or
+         (t.kind == tyObject and t.owner.kind == skType and
+          tfRefsAnonObj in t.owner.typ.flags):
```

Comment:
For the top level check wouldn't it make sense to check if the owning symbol is not an `skModule`, or would that be a problem for generic instantiations?

Reply:
Testing whether the owner is an `skModule` symbol wouldn't be able to detect the `Nested` type from `tseparate_hooks1.nim` (where the owner is a module symbol).

`sem` marks such types' symbols with the `sfGenSym` at the moment, so that they can be detected here. The only proper solution to this mess is getting rid of `sighashes`, really.